luarules/gadgets/unit_custom_weapons_cluster.lua

```diff
@@ -59,281 +58,258 @@ local cos   = math.cos
 
---------------------------------------------------------------------------------------------------------------
--- Initialize ------------------------------------------------------------------------------------------------
-
--- Information table for cluster weapons
+defaultSpawnTtl = defaultSpawnTtl * gameSpeed
 
 local spawnableTypes = {
-    Cannon          = true  ,
-    EMGCannon       = true  ,
-    Fire            = false , -- but possible
-    LightningCannon = false , -- but possible
-    MissileLauncher = false , -- but possible
+	Cannon          = true  ,
+	EMGCannon       = true  ,
 }
 
-local dataTable      = {} -- Info on each cluster weapon
-local wDefNamesToIDs = {} -- Says it on the tin
-
-for wdid, wdef in pairs(WeaponDefs) do
-    wDefNamesToIDs[wdef.name] = wdid
-
-    if wdef.customParams and wdef.customParams[customParamName] then
-        dataTable[wdid] = {}
-        dataTable[wdid].number  = tonumber(wdef.customParams.number) or defaultSpawnNum
-        dataTable[wdid].def     = wdef.customParams.def
-        dataTable[wdid].projDef = -1
-        dataTable[wdid].projTtl = defaultSpawnTtl
-        dataTable[wdid].projVel = defaultVelocity / GAME_SPEED
-
-        -- Enforce limits, eg the projectile count, at init.
-        dataTable[wdid].number  = min(dataTable[wdid].number, maxSpawnNumber)
-
-        -- When the cluster munition name isn't specified, search for the default.
-        if dataTable[wdid].def == nil then
-            local search = ''
-            for word in string.gmatch(wdef.name, '([^_]+)') do
-                search = search == '' and word or search .. '_' .. word
-                if UnitDefNames[search] ~= nil then
-                    dataTable[wdid].def = search .. '_' .. defaultSpawnDef
-                end
-            end
-            -- There's still the chance we haven't found anything, so:
-            if dataTable[wdid].def == nil then
-                Spring.Echo('[clustermun] [warn] Did not find cluster munition for weapon id ' .. wdid)
-                dataTable[wdid] = nil
-            end
-        end
-    end
+local clusterWeaponDefs = {}
+
+for unitDefName, unitDef in pairs(UnitDefNames) do
+	for _, weapon in pairs(unitDef.weapons) do
+		local weaponDefID, weaponDef = weapon.weaponDef, WeaponDefs[weapon.weaponDef]
+		local custom = weaponDef.customParams
+
+		if custom.cluster then
+			local clusterDefName = custom.cluster_def
+			local clusterCount = tonumber(custom.cluster_number) or defaultSpawnNum
+
+			if clusterCount < minSpawnNumber or clusterCount > maxSpawnNumber then
+				Spring.Log(gadget:GetInfo().name, LOG.WARNING, weaponDef.name .. ': cluster_count of ' .. clusterCount .. ', clamping to ' .. minSpawnNumber .. '-' .. maxSpawnNumber)
+				clusterCount = math.clamp(clusterCount, minSpawnNumber, maxSpawnNumber)
+			end
+
+			if clusterDefName then
+				if not WeaponDefNames[clusterDefName] then
+					-- Every weapon name contains its unit's name, per weapondefs_post.
```

Comment:
Not technically true but it's a convenient lie and maybe there aren't actually any counterexamples in BAR.

Reply:
He really means "every unit's weapon name." Preprocessing the `weapons` table inside the unit def prepends the unit's name.

Reply:
I know, but you can (but most likely don't) have weapons that don't follow that pattern since it only applies conditionally:
https://github.com/beyond-all-reason/Beyond-All-Reason/blob/1d5ad639958415db971dd9560df1857015bfbacb/gamedata/weapondefs_post.lua#L40-L41

Reply:
I'm torn on this, like with the default `cluster_number` above. On the one hand, single responsibility principle says all logic related to cluster weapons should be contained in the gadget, but on the other hand, to get expected engine default behaviour code needs to go in `gamedata`.

Reply:
You could make `LuaRules/Gadgtes/Cluster/` a folder, then have that as the "single" place that contains both the actual gadget and a file that isn't parsed by the gadget handler containing processing code to be included and ran by `gamedata/foodefs_post.lua`.

Reply:
Perhaps come up with an overview of how you'd ideally see it if engine wasn't a constraint and then we can work through it to see whether there are any limitations that can be overcome and/or turn out not to be engine limitations.

Reply:
There's really no way around it, defs processing must be done in the LuaParser environment, it can't be done in LuaRules. Of course, that brings up how do we support modding in the LuaParser environment, but that's a question for another time.

Reply:
Except that means defining default values outside of the gadgets 🤮 
I can't think of an elegant way to do this with the current architecture.

Reply:
> There's really no way around it, defs processing must be done in the LuaParser environment, it can't be done in LuaRules.

Yes, but gadgets already mix two environments (synced and unsynced luarules). So for example you could have
```diff
  if gadgetHandler:IsSyncedCode() then
    -- environment: synced luarules
- else
+ elseif gadgetHandler:IsUnsyncedCode() then
    -- environment: unsynced luarules
+ elseif gadgetHandler:NotActuallyGadgetHandlerButPostProcessing() then
+   -- environment: luaparser
+   -- for bla in pairs(DEFS.UnitDefs) do ...
end
```

Reply:
[Discord thread about added def tags and default values](https://discord.com/channels/549281623154229250/1347249345765969980)

Reply:
I'm torn on this, the place to implement it without duplicating the name prefix login is in weapondefs_post.lua, but it seems out of place to put any gadget-specific logic in there, it deals with global processing. Putting it in alldefs_post duplicates the logic, but in defs processing rather than in gadget.

Since there is nothing special that denotes the custom param as referring to a weapon def - it's just some arbitrary string - the special handling will have to go somewhere, the author could even do it in the unit def file itself.

Reply:
Implemented inside alldefs_post.lua